### ISSUE TITLE
Could not generate .h file for .so library

### ISSUE THREAD

**Issue by pavel-odintsov** [osrg/gobgp#374]
Hello!

I have checked master code because I want to pull updated C++ library.

But I could not build .so correctly:
```bash
cd /usr/src
git clone https://github.com/osrg/gobgp.git
cd gobgp/gobgp
go build -buildmode=c-shared -o libgobgp.so *.go
```

But I got only libgobgp.so and could not find libgobgp.h.

So even with old  libgobgp.h it do not work correctly:
```bash
undefined reference to `decode_path'
undefined reference to `serialize_path'
```

Could you help me?

**Comment by ishidawataru** [comment]
Hi

try `cd gobgp/gobgp/lib` instead of `cd gobgp/gobgp` !

**Comment by pavel-odintsov** [comment]
Nice! Thanks :)

**Comment by jeyamb** [comment]
In the recent "https://github.com/osrg/gobgp.git" , where the lib directory will be available ?

the lib directory is not found. when i modified the build.sh to pointing "cd ${GOBGP_PATH}/cmd/gobgp/" instead of "cd ${GOBGP_PATH}/cmd/gobgp/lib/" the "libgobgp.h" was not generated?  can you please help how to generate "libgobgp.h" in the recent version?

#!/bin/bash

GOBGP_PATH=${GOPATH}/src/github.com/osrg/gobgp

**cd ${GOBGP_PATH}/cmd/gobgp/**
go build -buildmode=c-shared -o libgobgp.so *.go
cd ${GOBGP_PATH}/tools/grpc/cpp
ln -sf ${GOBGP_PATH}/cmd/gobgp/libgobgp.h

**Comment by fujita** [comment]
https://github.com/osrg/gobgp/issues/1855

The next version doesn't support this feature. No need to use the binary format for the APIs.

**Comment by jeyamb** [comment]
But still its document in the same to way here 

https://github.com/osrg/gobgp/blob/master/docs/sources/grpc-client.md#cpp



**C++**
We use .so compilation with golang, please use only 1.5 or newer version of Go Lang.

'tools/grpc/cpp/gobgp_api_client.cc' shows an example for getting neighbor's information.

We provide 'tools/grpc/cpp/build.sh' to build this sample code. This script also generates stub codes and builds GoBGP shared library.

Let's build the sample code:

$ cd $GOPATH/src/github.com/osrg/gobgp/tools/grpc/cpp
$ bash build.sh
Let's run it
$ ./gobgp_api_client 172.18.0.2
BGP neighbor is: 10.0.0.2, remote AS: 1
	BGP version: 4, remote route ID
	BGP state = active, up for 0
	BGP OutQ = 0, Flops = 0
	Hold time is 0, keepalive interval is 0seconds
	Configured hold time is 90
BGP neighbor is: 10.0.0.3, remote AS: 1
	BGP version: 4, remote route ID
	BGP state = active, up for 0
	BGP OutQ = 0, Flops = 0
	Hold time is 0, keepalive interval is 0seconds
	Configured hold time is 90

**Comment by jeyamb** [comment]
Basically, i am developing a sample client application to Add and Del route for GoBGP. 
Is there any other latest documentation available to create c++ client application?

**Comment by jeyamb** [comment]
Thanks for your reply.


In the "https://github.com/osrg/gobgp/blob/master/docs/sources/grpc-client.md#cpp"  it was documented as mentioned below.


**C++**

We use .so compilation with golang, please use only 1.5 or newer version of Go Lang.

'tools/grpc/cpp/gobgp_api_client.cc' shows an example for getting neighbor's information.

We provide 'tools/grpc/cpp/build.sh' to build this sample code. This script also generates stub codes and builds GoBGP shared library.


When, i followed the same i end up with the following error

	[jbalasubramanian@GoBGP-Instance-1-beta.naw01.spdev.viasat.io ~/projects/src/github.com/osrg/gobgp/tools/grpc/cpp] (master)
	$ bash build.sh 
	g++ -std=c++11 -I/usr/local/include -pthread  -c -o gobgp_api_client.o gobgp_api_client.cc
	gobgp_api_client.cc:16:26: fatal error: libgobgp.h: No such file or directory
		 #include "libgobgp.h"
							  ^
	compilation terminated.
	make: *** [gobgp_api_client.o] Error 1

Since, the libgobgp.h is not generated, i am not able to proceed with the sample client application compilation. 


Basically, i am new to gobgp and trying to develop a sample client application to Add and Del route for GoBGP. 
Is this possible to proceed with developing an client application without "libgobgp.h"  header file?
Can you please help me to proceed further? Or share any documents that are available to develop a sample cpp application?

Also, if i develop an application with 1.33 do i need to revamp it after 2.x release ?

**Comment by jeyamb** [comment]
> No need to use the binary format for the APIs.

Can you please clarify following regarding compilation 

1. "The next version doesn't support this feature.". To have a common sync on this statement does the support is not available in next version, In the existing latest version 
   (goBGP 1.33) itself we are facing the problem. Is there any way to resolve this in this version?
   
2. "No need to use the binary format for the APIs.". If you mean the binary (libgobgp.so) format is removed. How the client application  
   will get the gobgpapi implementations? Is there any alternate mechanism available? 

Thanks and Regards
Jeya Murugan B

**Comment by fujita** [comment]
The old AddPath API uses the wire format contents of BGP path attributes.
https://github.com/osrg/gobgp/blob/d6c16da8988298e1098b50bb57af45004e0c146e/api/gobgp.proto#L554
libgobgp.so is for building the wire format contents of BGP attributes.

The new AddPath API uses the protobuf structures for BGP path attributes instead of the wire format contents.
https://github.com/osrg/gobgp/blob/master/api/attribute.proto

So libgobgp.so isn't necessary.

**Comment by jeyamb** [comment]
Thanks for you reply.

I have tried with the attributes present in the attribute.proto file to add route/path.  However, the  path information is not updated in gobgpd side.

The below is the addRoute implementation is which i have tried in my lab. I am suspecting somewhere packing the path attributes is not implemented correctly . Can you please help to resolve the issue.

        void addRoute (std::string sPrefix, std::string sNextHop)
        {
            std::cout <<"In addRoute \n";
            // Parameters to AddPath API
            gobgpapi::AddPathRequest request;
            ClientContext context;
            gobgpapi::AddPathResponse response;

            // Path info variable 
            gobgpapi::Path * current_path = new gobgpapi::Path;

            // Updating family info of current_path
            gobgpapi::Family * current_family = new gobgpapi::Family;
            current_family->set_afi(gobgpapi::Family::AFI_IP);
            current_path->set_allocated_family(current_family);

            // Updating nlri info for current_path
            google::protobuf::Any * current_nlri = new google::protobuf::Any;
            gobgpapi::IPAddressPrefix current_ipaddrprefix;
            current_ipaddrprefix.set_prefix(sPrefix);
            current_ipaddrprefix.set_prefix_len(32);
            current_nlri->PackFrom(current_ipaddrprefix);
            current_path->set_allocated_nlri(current_nlri);

            // Updating OriginAttribute info for current_path
            google::protobuf::Any* current_origin = current_path->add_pattrs();
            gobgpapi::OriginAttribute current_origin_t;
            current_origin_t.set_origin(0);
            current_origin->PackFrom(current_origin_t);

            // Updating NextHopAttribute info for current_path
            google::protobuf::Any* current_next_hop = current_path->add_pattrs();
            gobgpapi::NextHopAttribute current_next_hop_t;
            current_next_hop_t.set_next_hop(sNextHop);
            current_next_hop->PackFrom(current_next_hop_t);
            // Updating CommunitiesAttribute and UUID info for current_path
            google::protobuf::Any* current_communities = current_path->add_pattrs();
            gobgpapi::CommunitiesAttribute current_communities_t;
            current_communities_t.add_communities(100);
            current_communities->PackFrom(current_communities_t);

            // Populating the request attributes
            request.set_resource(gobgpapi::Resource::GLOBAL);
            request.set_allocated_path(current_path);

            stub_->AddPath(&context, request, &response);
    }

**Comment by pavel-odintsov** [comment]
It looks correct from my perspective. But I do not use AddPath logic.

**Comment by fujita** [comment]
@jeyamb I modified your code and pushed it as a C++ grpc example to add a route:
https://github.com/osrg/gobgp/blob/master/docs/sources/grpc-client.md#c
